java/generate-offline-map-with-local-basemap/src/main/java/com/esri/arcgisruntime/sample/generateofflinemapwithlocalbasemap/MainActivity.java

```diff
@@ -0,0 +1,384 @@
+      } else {
+        String error = "Error in generate offline map job: " + mGenerateOfflineMapJob.getError().getAdditionalMessage();
+        Toast.makeText(this, error, Toast.LENGTH_LONG).show();
+        Log.e(TAG, error);
+      }
+    });
+
+    // start the job
+    mGenerateOfflineMapJob.start();
+
+    if (findProgressDialogFragment() == null) {
+      ProgressDialogFragment progressDialogFragment = ProgressDialogFragment.newInstance(
+          getString(R.string.generate_offline_map_job_title),
+          getString(R.string.taking_map_offline_message),
+          getString(R.string.cancel)
+      );
+      progressDialogFragment.show(getSupportFragmentManager(), ProgressDialogFragment.class.getSimpleName());
+    }
+
+    // show the job's progress with the progress dialog
+    mGenerateOfflineMapJob.addProgressChangedListener(() -> {
+      if (findProgressDialogFragment() != null) {
+        findProgressDialogFragment().setProgress(mGenerateOfflineMapJob.getProgress());
+      }
+    });
+  }
+
+  /**
+   * Creates an alert notifying the user that a local basemap has been found on the device and asks whether the user
+   * wishes to use that basemap, rather than download one with the rest of the generate offline map job.
+   */
+  private void showLocalBasemapAlertDialog(String localBasemapFileName) {
+    if (getSupportFragmentManager().findFragmentByTag(LocalBasemapAlertDialogFragment.class.getSimpleName()) == null) {
+      LocalBasemapAlertDialogFragment localBasemapAlertFragment = LocalBasemapAlertDialogFragment.newInstance(
+          getString(R.string.local_basemap_found),
+          getString(R.string.local_basemap_found_message, localBasemapFileName),
+          getString(R.string.yes),
+          getString(R.string.no)
+      );
+      localBasemapAlertFragment
+          .show(getSupportFragmentManager(), LocalBasemapAlertDialogFragment.class.getSimpleName());
+    }
+  }
+
+  /**
+   * Callback from the local basemap alert dialog. Sets the reference basemap directory and calls generateOfflineMap().
+   */
+  @Override public void onPositiveClick() {
+    // set the directory of the local base map to the parameters
+    mGenerateOfflineMapParameters.setReferenceBasemapDirectory(mLocalBasemapDirectory);
+    // call generate offline map with parameters which now contain a reference basemap directory
+    generateOfflineMap();
+  }
+
+  /**
+   * Callback from the local basemap alert dialog. Leaves the reference basemap directory empty and calls
+   * generateOfflineMap().
+   */
+  @Override public void onNegativeClick() {
+    // call generate offline map with parameters which contain an empty string for reference basemap directory
+    generateOfflineMap();
+  }
+
+  /**
+   * Find and return the progress dialog fragment.
+   * @return the progress dialog fragment.
+   */
+  private ProgressDialogFragment findProgressDialogFragment() {
```

Comment:
This is used so we don't display two instances of the same `Fragment`. You could do this for the loading dialog too.

Reply:
Ahh ok. Makes sense. I know this is definitely more robust, but is there a way in this sample that two progress or alert dialogs could reasonably be created with the current UI? If not, I'm tempted to refactor away from the find___DialogFragment() type methods (rather than implement it for the AlertDialog fragment.

Reply:
I think if you spammed the trigger that created and commits the dialog, there's a chance it could happen and I can't remember off the top of my head but I think an exception is thrown if you try to re-add the same fragment with the same tag to the screen.

This method is handy for we want to update the progress UI, as we'll need to keep trying to find by tag.